### PR TITLE
Feature 157 and 112

### LINKED ISSUE #157: As a user I want to be prompted an error message when failing to login correctly during a change from guardian to citizen
> When providing a wrong password in order to change from a citizen to a guardian, no error message is prompted, and to my understanding this is a part of #138. We discussed earlier this week that the issue that my group is currently developing #112 and this user story require changes to the same code, and I am not 100% sure about what we agreed upon about this missing functionality. But I suggest that we either:
> 
> 1. Create a new issue regarding the missing functionality, that my group assigns itself to, and thus it will not be part of this PR.
>    or
> 2. The functionality is to implemented as part of this PR.

I think it should be a separate issue, since it has to be implemented in the classes where it is used, and not in the Auth Bloc. This makes it required that every bloc that has to login, which means that a general solution cannot be provided, but requires a specialized solution to work.

_Originally posted by @Tricky12321 in https://github.com/aau-giraf/weekplanner/pull/156#issuecomment-483597820_

### PR DESCRIPTION
**Pull request by User12837985123**
This closes #157 and #112.

### CODE REVIEW

**Comment by User12837985123** [comment]
@aau-giraf/sw609f19 you have been assigned this. Here's your checklist. I marked @Tricky12321 as the reviewer, but you might change it as you like.
Please use the following checklist. Feel free to check it off as you go along.
**Code Design**
- [ ] Is the code in the right place? (Both in terms of folder structure and class structure)
- [ ] Could this code, to the best of your knowledge, have reused existing code?
- [ ] Is the code over-engineered? Examples of over-engineering: Implemented behavior for future problems
- [ ] Does it introduce functionality that is not necessary for solving the problem?

**Code Readability and Maintainability**
- [ ] Are names meaningful and self-documenting?
- [ ] Is it understandable, by reading the code, what it does?
- [ ] Is the code simple and readable? i.e. it contains no “hack”, or obscure solution.
- [ ] Is it understandable, by reading the tests, what they do?
- [ ] Do the tests cover sensible cases? Both happy paths and exception paths?
- [ ] Do the tests cover the full functionality? i.e., will the tests fail if some of the requested functionality is missing?
- [ ] Are error messages in the code understandable?
- [ ] Are confusing sections of code covered by comments?

**Functionality**
For this part, open the app, and test:
- [ ] Are the features that the code claims to implement, actually implemented?
- [ ] Does it look like the code has bugs?
- [ ] Are all new screens reachable through in-app navigation?

**Points for Discussion and Reflection**
- [ ] Are UI messages, that the user might meet, actually helpful?
- [ ] Are there errors that are bad practice? Like, hard-coding a test database connection?
- [ ] Is there well written code in this? Remember to tell the developer!

**Comment by User12837985123** [comment]
@aau-giraf/sw610f19 you have been assigned this. Here's your checklist. I marked @Moesgaarda as the reviewer, but you might change it as you like.
Please use the following checklist. Feel free to check it off as you go along.
**Code Design**
- [x] Is the code in the right place? (Both in terms of folder structure and class structure)
- [x] Could this code, to the best of your knowledge, have reused existing code?
- [x] Is the code over-engineered? Examples of over-engineering: Implemented behavior for future problems
- [x] Does it introduce functionality that is not necessary for solving the problem?

**Code Readability and Maintainability**
- [x] Are names meaningful and self-documenting?
- [x] Is it understandable, by reading the code, what it does?
- [x] Is the code simple and readable? i.e. it contains no “hack”, or obscure solution.
- [x] Is it understandable, by reading the tests, what they do?
- [x] Do the tests cover sensible cases? Both happy paths and exception paths?
- [x] Do the tests cover the full functionality? i.e., will the tests fail if some of the requested functionality is missing?
- [x] Are error messages in the code understandable?
- [x] Are confusing sections of code covered by comments?

**Functionality**
For this part, open the app, and test:
- [x] Are the features that the code claims to implement, actually implemented?
- [x] Does it look like the code has bugs?
- [x] Are all new screens reachable through in-app navigation?

**Points for Discussion and Reflection**
- [x] Are UI messages, that the user might meet, actually helpful?
- [x] Are there errors that are bad practice? Like, hard-coding a test database connection?
- [x] Is there well written code in this? Remember to tell the developer!

File: lib/blocs/toolbar_bloc.dart [hunk 1]
```diff
@@ -376,9 +395,49 @@ class ToolbarBloc extends BlocBase {
...
+      //Pop the loading spinner.
+      Routes.pop(_currentContext);
+      showDialog<Center>(
+          barrierDismissible: false,
+          context: _currentContext,
+          builder: (BuildContext context) {
+            return const GirafNotifyDialog(
+                title: 'Fejl',
+                description: 'Forkert adgangskode',
+                key: Key('WrongUsernameOrPasswordDialog'));
```
Comment (Tekrus) [inline lib/blocs/toolbar_bloc.dart]:
```suggestion
                key: Key('WrongPasswordDialog'));
```

File: test/screens/weekplan_screen_test.dart [hunk 2]
```diff
@@ -106,7 +142,7 @@ void main() {
     api.week = weekApi;
     bloc = WeekplanBloc(api);
     authBloc = MockAuthBlock(api);
-    authBloc.setMode(WeekplanMode.guardian);
+    //authBloc.setMode(WeekplanMode.guardian);
```
Comment (Moesgaarda) [inline test/screens/weekplan_screen_test.dart]:
```suggestion

```

Comment (Tekrus) [inline test/screens/weekplan_screen_test.dart]:
```suggestion

```

File: lib/blocs/toolbar_bloc.dart (hunk 1 above)
Comment (Tekrus) [reply lib/blocs/toolbar_bloc.dart]: If this is included, remember to update the tests, it's a super minor thing tho

**Review by Tekrus — CHANGES_REQUESTED** [review]

**Review by Tekrus — DISMISSED** [review]

File: lib/blocs/auth_bloc.dart [hunk 3]
```diff
@@ -43,6 +50,18 @@ class AuthBloc extends BlocBase {
     });
   }
 
+  /// Authenticates the user only by password when signing-in from PopUp.
+  void authenticateFromPopUp(String username, String password) {
+    // Make sure the status for the upcoming
+    // login is false until proven otherwise
+    _api.account.login(username, password).take(1).listen((bool status) {
```
Comment (Tricky12321) [inline lib/blocs/auth_bloc.dart]: take(1) should be removed, it is not needed

File: test/widgets/giraf_app_bar_widget_test.dart [hunk 4]
```diff
@@ -45,13 +32,19 @@ class MockScreen extends StatelessWidget {
...
+            }));
   }
 }
 
 /// Used to retrieve the visibility widget wrapping the editbutton
 const String keyOfVisibilityForEdit = 'visibilityEditBtn';
+const String keyOfWrongUsernameOrPassword = 'WrongUsernameOrPasswordDialog';
+const String keyOfChangeToGuardian = 'IconChangeToGuardian';
+const String keyOfPasswordField = 'SwitchToGuardianPassword';
+const String keyOfConfirmButton = 'SwitchToGuardianSubmit';
```
Comment (Tricky12321) [inline test/widgets/giraf_app_bar_widget_test.dart]: I would argue to put these variables in to the Main function so they are not global.

**Review by Tricky12321 — CHANGES_REQUESTED** [review]
Small changes

File: test/widgets/giraf_app_bar_widget_test.dart (hunk 4 above)
Comment (User12837985123) [reply test/widgets/giraf_app_bar_widget_test.dart]: I can see they aren't even used anymore, so I will just delete them :D

File: lib/blocs/auth_bloc.dart (hunk 3 above)
Comment (User12837985123) [reply lib/blocs/auth_bloc.dart]: I am a bit interested. How can it then come that it is done in the authenticate method? Or should it also be removed there?

File: test/widgets/giraf_app_bar_widget_test.dart (hunk 4 above)
Comment (Tricky12321) [reply test/widgets/giraf_app_bar_widget_test.dart]: Great!

File: lib/blocs/auth_bloc.dart (hunk 3 above)
Comment (Tricky12321) [reply lib/blocs/auth_bloc.dart]: It should also be removed there, it is part of how the API used to work, and is no longer needed anywhere

**Review by Tricky12321 — APPROVED** [review]

**Review by Tekrus — APPROVED** [review]

File: lib/blocs/auth_bloc.dart (hunk 3 above)
Comment (Tekrus) [reply lib/blocs/auth_bloc.dart]: I've updated it in the authenticate method in PR #247 - seems like that was the only place still using the .take(1)

**Review by Eduardsen — APPROVED** [review]

**Pull request merged**